### PR TITLE
Improvements to javametrics injection + tests

### PR DESCRIPTION
**Pull request by mattcolegate**
Signed-off-by: Matthew Colegate <colegate@uk.ibm.com>

This PR contains updates to MetricsService that allow Spring applications to be injected with Metrics code, as well as making the injection process for Java applications more surgical. Tests have been updated and new ones written to maintain coverage.

### CODE REVIEW

**Comment by mattcolegate** [comment]
@rwalle61 would also appreciate your input

File: test/resources/metricsService/spring/Application.java/package3/with collector.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package application3;
```
Comment (rwalle61) [inline test/resources/metricsService/spring/Application.java/package3/with collector.java]: what's the difference between package3 and packages 1 and 2?

File: src/pfe/portal/modules/MetricsService.js [hunk 2]
```diff
@@ -214,129 +250,210 @@ function getNewContentsOfPomXml(originalContents) {
...
+  }
+  
+  
+  let metricsCollectorBuildPlugin = originalBuildPlugins[mavenDepPluginIndex];
+  if (metricsCollectorBuildPlugin.executions) {
+    metricsCollectorBuildPlugin.executions[0].execution = metricsCollectorBuildPlugin.executions[0].execution.concat(newExecutionArray);
+  } else {
+    metricsCollectorBuildPlugin.executions = [ { execution: newExecutionArray } ];
+  }
+  let newBuildPlugins = originalBuildPlugins;
```
Comment (rwalle61) [inline src/pfe/portal/modules/MetricsService.js]: overwriting `originalBuildPlugins` is dangeous, we could do `const newBuildPlugins = deepCopy(originalBuildPlugins)`?

File: src/pfe/portal/modules/MetricsService.js [hunk 3]
```diff
@@ -214,129 +250,210 @@ function getNewContentsOfPomXml(originalContents) {
...
+          execution: newExecutionArray,
+        }
+      ]
+    };
+    const newBuildPlugins = originalBuildPlugins.concat(metricsCollectorBuildPlugin);
+    return newBuildPlugins;
+  }
+  
+  
+  let metricsCollectorBuildPlugin = originalBuildPlugins[mavenDepPluginIndex];
```
Comment (rwalle61) [inline src/pfe/portal/modules/MetricsService.js]: Was this already defined above (with a different name?)

File: src/pfe/portal/modules/MetricsService.js [hunk 4]
```diff
@@ -122,19 +122,33 @@ async function injectMetricsCollectorIntoApplicationJava(pathToApplicationJava)
...
-
-  const indexOfSpringBootApplication = splitOriginalContents.findIndex(line => line === springAppAnnotation);
-  const metricsCollectorAnnotation = `@ComponentScan(basePackages = {"${packageName}", "com.ibm.javametrics.spring"})`;
-
+  // metrics collector not present, so...
   let newApplicationJava = deepClone(splitOriginalContents);
-  newApplicationJava.splice(indexOfSpringBootApplication + 1, 0, metricsCollectorAnnotation);
+  if (componentScanPresent != -1) {
+    // alter what's already in there
+    newApplicationJava[componentScanPresent] = newApplicationJava[componentScanPresent].replace('{', '{"com.ibm.javametrics.codewind.spring", ');
```
Comment (rwalle61) [inline src/pfe/portal/modules/MetricsService.js]: as `componentScanPresent` isn't a boolean nor used like one, should it have a different name?

File: test/src/unit/modules/MetricsService.test.js [hunk 5]
```diff
@@ -24,7 +24,7 @@ chai.should();
 
 const deepClone = (obj) => JSON.parse(JSON.stringify(obj));
 
-describe('MetricsService.js', () => {
+describe.only('MetricsService.js', () => {
```
Comment (rwalle61) [inline test/src/unit/modules/MetricsService.test.js]: eslint on your vscode should catch this

File: test/src/unit/modules/MetricsService.test.js [hunk 6]
```diff
@@ -102,7 +102,7 @@ describe('MetricsService.js', () => {
         expectedPomXmlForSpring = await readXml(pathToExpectedPomXmlForSpring);
 
         // To debug these parsed PomXmls
-        // const util = require('util');
+        //const util = require('util');
```
Comment (rwalle61) [inline test/src/unit/modules/MetricsService.test.js]: nit

File: test/src/unit/modules/MetricsService.test.js [hunk 7]
```diff
@@ -300,6 +300,15 @@ describe('MetricsService.js', () => {
             });
         });
 
+        describe('getNewContentsOfJvmOptions(existingContents)', () => {
+            it('returns a string representing the contents of jvm.options after injection - no change', () => {
```
Comment (rwalle61) [inline test/src/unit/modules/MetricsService.test.js]:
```suggestion
            it('returns the existing contents when they already contain our metrics collector', () => {
```

File: test/src/unit/modules/MetricsService.test.js [hunk 8]
```diff
@@ -536,7 +577,25 @@ describe('MetricsService.js', () => {
...
                 });
+                describe(testName, () => { // eslint-disable-line no-loop-func
+                    it('returns the object representing an Application.java injected with metrics collector with no changes', () => {
+                        const funcToTest = metricsService.__get__('getNewContentsOfApplicationJava');
+                        const output = funcToTest(test.expectedApplicationJava);
+                        output.should.deep.equalInAnyOrder(test.expectedApplicationJava);
+                    });
+                });
             }
+            const test3 = {
```
Comment (rwalle61) [inline test/src/unit/modules/MetricsService.test.js]: do these have to separated from the above tests?

**Comment by mattcolegate** [comment]
Superseded by #1364

**Pull request closed without merging**